# Q4 Apply best model on the test data

Test data path = /content/drive/MyDrive/DL_DATA_iNat/val


In [5]:
import os
import math
import numpy as np
import matplotlib.pyplot as plt
import keras
import tensorflow as tf
from tensorflow.keras.preprocessing import image
from tensorflow.keras.models import Model
import matplotlib.gridspec as gridspec
from PIL import Image
from tensorflow.keras.preprocessing.image import ImageDataGenerator

In [2]:
#mount drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
myBestModel = keras.models.load_model('drive/MyDrive/DL_DATA_iNat/model-best.h5')

In [9]:
myBestModel.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 227, 227, 32)      896       
                                                                 
 batch_normalization (BatchN  (None, 227, 227, 32)     128       
 ormalization)                                                   
                                                                 
 activation (Activation)     (None, 227, 227, 32)      0         
                                                                 
 max_pooling2d (MaxPooling2D  (None, 113, 113, 32)     0         
 )                                                               
                                                                 
 conv2d_1 (Conv2D)           (None, 113, 113, 64)      18496     
                                                                 
 batch_normalization_1 (Batc  (None, 113, 113, 64)     2

### Q4 - (a) -Use the best model and report the accuracy on the test set. 

In [6]:
BATCH_SIZE = 32
test_datagen = ImageDataGenerator(rescale=1./255)
test_data = test_datagen.flow_from_directory(
  '/content/drive/MyDrive/DL_DATA_iNat/val',
  target_size=(227,227),
  batch_size=BATCH_SIZE,
  class_mode='categorical')

Found 2009 images belonging to 10 classes.


In [8]:
myBestModel.evaluate(test_data,steps= test_data.n/test_data.batch_size)

62/62 [==============================] - 456s 7s/step - loss: 1.8102 - accuracy: 0.3758


[1.8101873397827148, 0.37580886483192444]

for best model on test data:
{Loss: **1.8102**} and {Accuracy: **0.3758**}

###Q4 - (b) - 10 x 3 grid containing sample images from the test data and predictions made by best model

In [ ]:
#Get all the classes from the folder (folder name is val but it is testing data, validation data was obtained by spliting the train data)
classes=[]
for fold_name in np.sort(os.listdir('/content/drive/MyDrive/DL_DATA_iNat/val')):
  classes.append(fold_name)
print(classes)

In [ ]:
samp_imgs = []
test_path = '/content/drive/MyDrive/DL_DATA_iNat/val'
labels = []
# -- total 30 images and 10 classes..
# -- Taking 3 random images from each class..
for i in range(10):
    files = os.listdir(test_path+classes[i])
    indices = np.random.choice(len(files),3)
    for j in range(3):
        pic = Image.open(test_path+classes[i]+'/'+files[indices[j]])
        pic = pic.resize((227,227))
        samp_imgs.append(np.asarray(pic)*1./255)
        labels.append(i)
        
samp_imgs = np.array(samp_imgs)